src/Microsoft.Extensions.WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,309 @@
+        /// <returns>A <see cref="Task"/> that completes when the message has been written to the outbound stream.</returns>
+        // TODO: De-taskify this to allow consumers to create their own awaiter.
+        public Task SendAsync(WebSocketFrame message, CancellationToken cancellationToken)
+        {
+            // Calculate the size of the frame
+            // Base header size is 2 bytes.
+            var headerSize = 2;
+            if (message.Payload.Buffer.Length > ushort.MaxValue)
+            {
+                // We're going to need an 8-byte length
+                headerSize += 8;
+            }
+            else if (message.Payload.Buffer.Length > 125)
+            {
+                // We're going to need a 2-byte length
+                headerSize += 2;
+            }
+            if(_masked)
+            {
+                headerSize += 4;
+            }
+
+            // Allocate a buffer
+            var buffer = _outbound.Alloc(minimumSize: headerSize);
+            if (buffer.Memory.Length < headerSize)
+            {
+                throw new InvalidOperationException("Couldn't allocate enough data from the channel to write the header");
+            }
+
+            // Write the opcode and FIN flag
+            var opcodeByte = (byte)((int)message.Opcode << 4);
+            if (message.EndOfMessage)
+            {
+                opcodeByte |= 1;
+            }
+            buffer.WriteBigEndian(opcodeByte);
+
+            // Write the length and mask flag
+            var maskingByte = 0; // TODO: Masking flag goes here
+
+            if (message.Payload.Buffer.Length > ushort.MaxValue)
+            {
+                buffer.WriteBigEndian((byte)(0xFE | maskingByte));
+
+                // 8-byte length
+                buffer.WriteBigEndian((ulong)message.Payload.Buffer.Length);
+            }
+            else if (message.Payload.Buffer.Length > 125)
+            {
+                buffer.WriteBigEndian((byte)(0xFC | maskingByte));
+
+                // 2-byte length
+                buffer.WriteBigEndian((ushort)message.Payload.Buffer.Length);
+            }
+            else
+            {
+                // 1-byte length
+                buffer.WriteBigEndian((byte)((message.Payload.Buffer.Length << 1) | maskingByte));
+            }
+
+            if(_masked)
+            {
+                WriteMaskingKey(ref buffer);
+            }
+
+            // Append the payload (this will copy the actual buffer struct, but NOT the underlying data)
+            var payload = message.Payload.Buffer;
+            buffer.Append(ref payload);
+
+            // Flush
+            return buffer.FlushAsync();
+        }
+
+        private void WriteMaskingKey(ref WritableBuffer buffer)
+        {
+            // NOOOOOOOOOOOO! Have to allocate :(. Can't stackalloc... RandomNumberGenerator only takes a 'byte[]' (not even ArraySemgent or the like)
```

Comment:
I'd like to keep this as thread-safe as possible. Buuut in theory it wouldn't be too bad to require that only one thread is performing write operations and one thread is performing read operations. And this would only be used by the writer.

@davidfowl How thread-safe are channels around multiple writes? I'm guessing there's nothing to defend against tearing the data (other than pre-allocating the whole buffer up front) so maybe my thread-safety concerns are moot and we can just allocate a single buffer for this...